### PR TITLE
Allow scala-native to be used with FreeBSD

### PR DESCRIPTION
**Pull request by alexdupre**

### CODE REVIEW

File: javalib/src/main/scala/java/net/Socket.scala
Comment (LeeTibbert) [inline javalib/src/main/scala/java/net/Socket.scala:251]: @alexdupre Wonderful to see this contribution.
I am an individual contributor, not a Scala Center person.
My background is in network programming (implementing
TCP, UDP, etc in a commercial O/S kernel.)

When I saw this changes, I immediately tweaked.  After a while
studying the change, the Open Group Specification, Java JVM practice 
(not source), and the underlying SN PlainSocketImpl.scala, I believe
this change will not have a negative impact on existing code. But then
you know that to begin with.  

As you are probably well aware. socket code is particularly fragile,
especially since it is hard to test if any given change is innocuous
and end users get cranky when their years old application stops working
after an upgrade.  Forgive me if I am preaching to the choir (saying something that is well known and already believed.)

I will review the rest of this PR as soon as I can.  If I do not get there
before Scala Center review, I do not want to hold this PR up.

File: posixlib/src/main/resources/scala-native/time.c
Comment (LeeTibbert) [inline posixlib/src/main/resources/scala-native/time.c:97]: I am probably in the minority and would have to check the linux
practice, but I have found it easier to follow (fewer gymnastics) 
the positive `defined` rather than the negative `!defined`.  That is
```
#if defined(__FreeBSD__)
           return 0;
#else 
            return timezone
#endif
```
Not only is it easier at present to see what is happening in the __FreeBSD__
case, it is also easier to know where the code goes when someone
down the line wants to add __SomeWierdBehavior__.

I think there are at least two cases of this idiom here.

You might want to see what someone from Scala Center  so that
my possibly minority position does not cause you to waste effort.

I mention it here for consideration for other reviewers. We all are going
to have to live with and maintain this code for a long time. 

I have not checked with what prior art in that file, if any, does.

File: unit-tests/src/test/scala/java/lang/ProcessTest.scala
Comment (LeeTibbert) [inline unit-tests/src/test/scala/java/lang/ProcessTest.scala:109]: Sorry for my lack of understanding.  What happens if, Heresy!, the shell is not bash?  Has that case been tested?  Should the name be, say, `shellExtraPaths`

I, for one, have never run the build on anything but bash and am
not sure if that is a requirement outside of `scripted`.

File: unit-tests/src/test/scala/java/lang/ProcessTest.scala
Comment (LeeTibbert) [inline unit-tests/src/test/scala/java/lang/ProcessTest.scala:130]: See comment re: shell   above,

File: unit-tests/src/test/scala/java/lang/ProcessUtils.scala
Comment (LeeTibbert) [inline unit-tests/src/test/scala/java/lang/ProcessUtils.scala:13]: Given that we are well off into the land of multiple distributions, it
iight be nice to have some comment/breadcrumg for how this apparently mongrel lot was painstakingly derived. 

Not that I am advocating it/them but I do not see a `sbin`. I also do not
see any `snap`.   

On Ubuntu (20.10) before I muck with the PATH for my process, I see
something like `/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin`.  Other distributions may have something else.

Thank you for considering my concern and feel free to let me know
that I am off base and lack comprehension.

File: unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala
Comment (LeeTibbert) [inline unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala:173]: I'll bight, why the change?  Seems to be a change by 2. A bit being set?

File: unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala
Comment (LeeTibbert) [inline unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:42]: Again, I am probably in the minority.  I dislike `assumeFalse` 
__intensely__ because it throws scary looking garbage into the 
log file. The garbage looks like an error, but is a bit of 'known error'
lore which must be re-parsed and re-learned after one has been
away for a month or so.

To me, a simple `if` which executes only if not FreeBSD conveys the
intent and does not generate `known error` garbage.  There is nothing
to warn about, the test is executing successfully on both FreeBSD and,
say, Ubuntu.  It is just on the former, the code sees that it has nothing to
do. There is no exceptional (with a small e) code flow to notify 
someone reading the log files about.

I spend a __lot__ of time reading SN logs, and so am somewhat sensitive
about wasting time skipping over "known errors' that I used to know.

Again, I may be in the minority, so you should probably see what a 
Scala Center person advises before making any change.

Comment (LeeTibbert) [reply unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:42]: And yes, I would leave a one line comment "FreeBSD does not have the 'timezone' variable".  Some poor soul down the line trying to read and
comprehend this stuff will sing your praises (which is why, I assume,
you put such a nice comment in the assumeFalse).

File: javalib/src/main/scala/java/net/Socket.scala
Comment (LeeTibbert) [inline javalib/src/main/scala/java/net/Socket.scala:249]: This PR does not touch this line, but it would be a mercy if a one line comment
were to be placed just above it explaining Java 8 specifies the maximum
value as "The maximum timeout value is platform specific" but actual
practice is to limit to 65545.   It took me a while to figure out this magic
number (slow child) and others tracing the code may have the same 
difficulty.  A breadcrumb would save others effort.

File: unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala
Comment (ekrich) [reply unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala:173]: A couple of comments on this change.
1. The results where generated on macOS but work on Linux as well.
2. Currently we are only doing CI on Linux but I think we need to consider supporting macOS and BSD if we can.
3. If looks like there is a small additional error on this test so maybe we can add something like the following:
```Scala
def isAlmostEqual(act: Float, exp: Float, ulpFactor: Int): Boolean = ...
// replace 2 in the function with ulpFactor
// Also add the same with a default to the following and make appropriate changes
def assertEqualsComplexF(act: Ptr[CFloatComplex],
                           exp: Ptr[CFloatComplex], ulpFactor = 2): Unit =...
// pass the ulpFactor through
```

Then for the test in question you can try `3` and leave a comment that BSD is off by three compared to macOS and Linux off by 2.

Comment (alexdupre) [reply unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala:173]: From what you say I understand that the literal numbers were generated as copied as-is from macOS. There is an identical test with `Cpow` that uses `double` instead of `float`, that I assume produces more accurated results. What I did was to take the complex double and convert it to a complex float. This is how I generated the number, it's not the result from FreeBSD, so I think it's a more accurate number (supported by the fact that  the CI on Linux didn't break). If macOS is fine too I don't think we need to change `2` with `3`.

File: posixlib/src/main/resources/scala-native/time.c
Comment (alexdupre) [reply posixlib/src/main/resources/scala-native/time.c:97]: If it were up to me I would have written it in the positive form as you suggested, but then I thought that perhaps the project prefers to have the "normal" code first and the "exception" later.
This comments applies to all other points where you found a negative form as well.

File: javalib/src/main/scala/java/net/Socket.scala
Comment (alexdupre) [reply javalib/src/main/scala/java/net/Socket.scala]: For reference, this is the code the breaks (and throws a SocketException) when the value is negative: https://github.com/freebsd/freebsd-src/blob/098dbd7ff7f3da9dda03802cdb2d8755f816eada/sys/kern/uipc_socket.c#L3017-L3022

And this is the Java native implementation that sets 0 when SO_LINGER is disabled: https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/master/jdk/src/solaris/native/java/net/PlainSocketImpl.c#L944-L951

But thanks to your comment I found out that my initial patch wasn't fully correct, and I also found another issue in the code. I've just pushed the final fix.

File: javalib/src/main/scala/java/net/Socket.scala
Comment (alexdupre) [reply javalib/src/main/scala/java/net/Socket.scala]: If you look at the linked FreeBSD source code above there is a check that the value is not > USHORT_MAX, so 65535 seems a safer max value.

File: unit-tests/src/test/scala/java/lang/ProcessTest.scala
Comment (alexdupre) [reply unit-tests/src/test/scala/java/lang/ProcessTest.scala:109]: This is not related to the shell you are using. This test is launching `echo.sh` and that is a bash script that previously had a fixed `/bin/bash` in the shebang line. All other scripts work with plain `/bin/sh`, this is the only one that requires bash, and it may be in a different path on other platforms. If in future it'll be added a test that requires zsh (I highly doubt) I guess we can change the name later.

File: unit-tests/src/test/scala/java/lang/ProcessUtils.scala
Comment (alexdupre) [reply unit-tests/src/test/scala/java/lang/ProcessUtils.scala:13]: As said above, this is just for bash at the moment. AFAIK Linux distributions have it in `/bin` and BSD variants in `/usr/local/bin`, I don't think shells are usually installed in `/sbin` or other paths.

File: unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala
Comment (alexdupre) [reply unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:42]: I'm not sure to agree on this one. I've added the `assume` on purpose exactly to not hide that FreeBSD lacks the timezone variable. The warning will show up only on FreeBSD and makes clear that you cannot rely on the timezone variable in your code if it's run on FreeBSD.

File: unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala
Comment (ekrich) [reply unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala:173]: That is fine if it works on macOS if you put a comment with the "// old value macOS, new value changed for BSD". The code generation is below so it can be repeated and studied - not sure if it is the best way or not.

The double code uses a `ulpFactor` of 1 or just `epsilon` vs 2 for float.

File: unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala
Comment (ekrich) [inline unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:44]: Rather than repeating this in the tests, I think we can add it to `Platform.scala` in the test `utils` directory. 

Maybe `isFreeBSD`?.

Just out of curiosity is this only FreeBSD or could we expect the other variants to work as well?

File: unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala
Comment (alexdupre) [reply unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala:173]: I've run the code. These are the FreeBSD values:
```
cpowf   -> toFloat(0x3e8c4421), toFloat(0x3f156d69)
cpow    -> toDouble("3fd18884016cf327"), toDouble("3fe2adad36b098aa")
```
So as you can see the double values are the same, while the float are different from both the original ones and the patched ones. As I said, the float value has been generated by converting the double value, so it should be more accurate than any direct output of `cpowf`. I cannot test it on macOS.

File: unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala
Comment (alexdupre) [reply unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:44]: You can expect it for other variants as well (OpenBSD, etc.)

File: unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala
Comment (alexdupre) [reply unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala:175]: It should work on macOS, too, the `assertEqualsComplexF` function returns `true` when I pass the original numbers.

File: javalib/src/main/scala/java/net/Socket.scala
Comment (alexdupre) [reply javalib/src/main/scala/java/net/Socket.scala:250]: Probably it was a typo, I've added a comment.

**Comment by LeeTibbert** [comment]
A note on CI testing.

I have been keeping this PR and how to do Continuous Integration of it on
FreeBSD in mind as I have been pecking away at restoring running CI
on macOS.  

Running this PR on Linux CI shows that it introduces no harm on Linux.
However, it does nothing to increase confidence that the code runs
as intended on FreeBSD.

From my studies for macOS GitHub Actions for CI on macOS,
I have read that a GitHub Action running on FreeBSD does not come
"out-of-the-box" but that a workflow can be set up to create
such an environment.

I am trying to bring macOS CI for SN to completion, so I have
not spent time on FreeBSD CI, beyond developing the belief
that it can be done with reasonable effort.

Once macOS SN CI is settled down and behaving, I hope
to take a run at creating a separate workflow for FreeBSD SN CI.
Like the macOS workflow, the FreeBSD workflow would probably
be private for an incubation period.

While a FreeBSD CI workflow would be nice, I thing that lack of
such should not stand in the way of this PR being merged.  

There is much to learn from exercising FreeBSD SN while a 
proper CI is being developed.  


PS:  Just saw the probably ancient "request for review".  I will try
        to loop back this weekend when I am likely to have the quiet,
        focused time this PR deserves.

**Comment by LeeTibbert** [comment]
@AlexandreDubray  When you have some time, could you please do a git rebase?
                                  A conflicting file has crept in.

**Comment by alexdupre** [comment]
@LeeTibbert re-based and fixed a new test.

File: unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala
Comment (LeeTibbert) [inline unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:291]: In the last two weeks or so there has been an update to TimeTest.scala
to allow it to additionally run on macOS.  TimeTest, and this section in
particular, is a recurring pain.

Do you know if strptime() on FreeBSD  either:
1) rejects %Z (implied posix behavior).
2) parses %Z and does nothing (gnu linux behavior)
3) parses %Z and sets C fields (macOS behavior)

If #1, then my scars tell me putting the "if (Platform.isFreeBSD)" at
 the beginning of the SN mainline TimeTest.scala test and skipping
 the entire thing is the way to go.  

If either  #2 or #3, the current SN mainline test should work, as is,
and not require any special FreeBSD handling. (Unless I am missing
something, I have been known to do that.)

If there is some other behavior, please let me know (or, I suggest, just
 bypass as in case #1 above).

Does this help?

**Comment by LeeTibbert** [comment]
@AlexandreDubray Thank you for rebase. Nice to see all green CI.

File: unit-tests/src/test/scala/java/lang/ProcessTest.scala
Comment (ekrich) [reply unit-tests/src/test/scala/java/lang/ProcessTest.scala:109]: Can `echo.sh` be changed to use `/bin/sh` rather than these extra changes?

File: unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala
Comment (ekrich) [reply unit-tests/src/test/scala/scala/scalanative/libc/CComplexTest.scala:175]: ok, that sounds good.

File: unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala
Comment (ekrich) [reply unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:49]: Yes, I believe this is how we use `assume`, precisely to highlight the assumptions for test coverage.

**Comment by alexdupre** [comment]
> @AlexandreDubray Thank you for rebase. Nice to see all green CI.

BTW, I'm *not* Alexandre Dubray :-)

File: unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala
Comment (alexdupre) [reply unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:291]: As I added in the comment, FreeBSD parses %Z and use it to create the equivalent time in the local timezone. This means that if the string contains GMT and my local timezone is CET (or CEST, given that Mar 31 is in DST), the hour check fails because 16 != 14.

Comment (LeeTibbert) [reply unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:291]: @alexdupre 

1) Sorry about getting your name wrong, as the actor famously said
    to the reporter: Print anything you want about me but spell my name right!

2) Given your info about %Z parsing, I will go back to the comment you mention.
    I have spent __way__ too much of my life on this test over the past two years
    or more. 

    The problematic test is not really about strptime() but about showing that
    SN is using structure copy-out/copy-in with its "short" POSIX 'struct tm'
     declaration.   

     I think that test is passing on macOS CI because the local time on that
     system is UTC.  That, plus what you are seeing means that at least
     the hours & minutes part of the current test are __wrong__ in the wild.
     I will have to PR that. I will probably wait until after this PR is merged.
     Thank you for detecting that flaw.
     
Reviewers:  I believe fixing the pre-exiting TimeTest.scala flaw is on me and should
                   not hold up this PR.
                   this PR.

Comment (alexdupre) [reply unit-tests/src/test/scala/scala/scalanative/posix/TimeTest.scala:291]: Yes, surely if the system is in UTC the issue cannot be seen. I don't know what are the linux and macos behaviors, the linux man page is not clear enough to me, while on FreeBSD it clearly says that `The resulting values will be relative to local time`, and I can confirm it. But as you said, the test is not really about strptime.

File: unit-tests/src/test/scala/java/lang/ProcessTest.scala
Comment (alexdupre) [reply unit-tests/src/test/scala/java/lang/ProcessTest.scala]: Done

**Comment by LeeTibbert** [comment]
Reviewers:

This PR looks good to me (LGTM) as far as safety.

It has revealed and works around a bug in TimeTest. I will
submit a subsequent PR to remove that bug after this PR
merges.

I have run the next-to-last version of this PR twice on private, incubator, macOS PR.
There were no errors attributable to it.  

I can not give independent validation of efficacy at the present.
I am working on creating a private, GitHub Actions incubator CI environment
 where SN is built and the tests run on FreeBSD.

That effort will take weeks, not days. I believe that it should not hold up
the merge of this PR.

**Comment by ekrich** [comment]
@alexdupre I think it would be a really good idea to add a section in the docs on how to setup FreeBSD to run with Scala Native. You might want to also include the minimum version or the version of FreeBSD that this should work on.

I am hoping too that you will continue tracking the project to keep BSD support going into the future 😄

**Comment by LeeTibbert** [comment]
There is an Open PR which describes how to setup and run the Sphinx doc build software
for Scala Native.  Scala Native Documentation uses Rich Structured Text (.rst) files.

To reduce the overhead/burden of submitting a PR, if you want to punt editing .rst
files and then building & displaying them in html, you can send me a plain text
file of what you would like to say and I will add the .rst tags and do the trial build.

Always nice to have someone else who can update the documentation, so I 
do not want to discourage you if you are interested in doc issues.  My offer
of help is to make a PR "do-able", so that contributors continue to contribute...

**Comment by ekrich** [comment]
Sure that would satisfy my request.

**Review by ekrich — APPROVED** [review]
LGTM, pending some user documentation volunteered by @LeeTibbert

**Comment by LeeTibbert** [comment]
Hot off the presses.  This is meant in __support__ of this PR, not to hold it up.

I have now run Scala Native `test-runtime` on a very preliminary, private shadow
FreeBSD CI.  That is, this PR running on FreeBSD.   

The build phase succeds, but there are some test failures.  I need to
study the tests involved.

```
[info] Test run finished: 0 failed, 0 ignored, 8 total, 36.627s
[error] Failed: Total 4164, Failed 6, Errors 0, Passed 4100, Skipped 58
[error] Failed tests:
[error] 	java.util.DefaultFormatterTest
[error] 	java.io.FileOutputStreamTest
[error] (tests / Test / test) sbt.TestsFailedException: Tests unsuccessful
```
I have not tried `test-tools` or `test-scripted`.   The FreeBSD message
in TimeTest.scala happened.

I am out (well past) out of time for today.  Will peck away at this as I get time.

**Comment by alexdupre** [comment]
> [error] 	java.util.DefaultFormatterTest
> [error] 	java.io.FileOutputStreamTest

FWIW both these tests run fine here (as the entire suite), but I don't exclude there may be differences in the environment that can cause it (e.g. filesystem, jdk, charset, lang/locale differences, etc.), like the timezone difference that we spotted in the other test

**Comment by ekrich** [comment]
Ok, I am good with this PR. Pending @LeeTibbert and @WojciechMazur. I put some comments  on the POSIX signal PR.

**Comment by LeeTibbert** [comment]
re: new FreeBSD documentation.  @alexdupre I hope to be your first customer. That is, I 
plan on using what your wrote, re-versions, etc.  to inform my Incubator CI on FreeBSD
efforts.  Then we can run periodic community CI runs to detect if any PRs cause
problems in FreeBSD (and to have a working example of a SN FreeBSD environment
to show people).

**Review by LeeTibbert — COMMENTED** [review]
LGTM, Thank you for all the hard work.

In support of this work: 

1) I have been running, on FreeBSD,  with these changes as I
have been working on and fixing bugs in SN mainline
repository revealed by this PR.  All good so far.

2) My private GitHubAction to do CI runs using the
    FreeBSD operating system, used above, is not ready for SN
    mainline, but I will continue to peck away at it,
    My goal is that merges to SN mainline get exercised on
    macOS and FreeBSD in community runs.

File: posixlib/src/main/resources/scala-native/time.c
Comment (WojciechMazur) [reply posixlib/src/main/resources/scala-native/time.c:165]: Let's change it to the positive condition as LeeTibbert suggested.

File: posixlib/src/main/resources/scala-native/errno.c
Comment (WojciechMazur) [inline posixlib/src/main/resources/scala-native/errno.c:87]: Can we move pre-processor conditions to the actual body of the function? It's the convention used in most of the other sources and in my opinion it's more readable, even if it's a simple oneliner. 

```suggestion

int scalanative_enodata() { 
#ifdef ENODATA
    return ENODATA;
#else
    return 0; 
#endif
}

```

```

File: posixlib/src/main/resources/scala-native/termios.c
Comment (WojciechMazur) [inline posixlib/src/main/resources/scala-native/termios.c:48]: Same as in `ernno.c`, can we move conditions to the body of the functions.

**Review by WojciechMazur — CHANGES_REQUESTED** [review]
Looks good, I've pointed only some issues about pre-processor conditions.

File: posixlib/src/main/resources/scala-native/termios.c
Comment (alexdupre) [reply posixlib/src/main/resources/scala-native/termios.c:49]: Done

**Comment by alexdupre** [comment]
Something broke in the CI script: 

```CLANG_FORMAT_PATH='/usr/lib/llvm-6.0/bin/clang-format' does not exist or is not executable```

**Comment by ekrich** [comment]
Confirmed the issue - https://github.com/scala-native/scala-native/issues/2238

**Review by WojciechMazur — APPROVED** [review]

**Pull request merged**
